### DISCUSSION TITLE
Merge two TIFFs in one with alpha channel

### DISCUSSION THREAD

**Discussion by cesqo** [ImageMagick/ImageMagick#5860] [Consulting]
Hello everyone.

I have a request. I'm not an expert with command lines :( 

I try to explain my need: I have two grayscale tiffs files (transparent) that I want to merge into one file. One of those it should remain in the gray channel, and the other one I have to put into an alpha channel in the same file as I said.

If anyone can help me with the command line I really appreciate that 

Best,
Francesco

**Comment by fmw42** [comment]
What is your IM version and platform/OS?  You will need IM 7 to what you want.

Please clarify where you want to gray channels to go and where you want the alpha channels to go from your two tiff files.  Do you merge the two gray channels, do you use just one of them. Do you combine the alpha channels from the tiff.  Are the transparencies in the tiffs in the background or the alpha channels?

Please post your two tiff files.  Zip them and then you should be able to upload them here. Or post them to some free hosting service that won't change the formats and put the URLs here

**Reply by cesqo** [reply]
Hi @fmw42 

My OS is Mac 12.3.1 and ImageMagick 7.1.0-53 Q16-HDRI aarch64 20646

Yes I attached you the two tiffs that I had (file-print.tiff and file-alpha.tiff) and the result that I want (file-output.tiff) here ( https://www.dropbox.com/t/Hg1O94nTO6mU7XkG )

I want to merge the two (file-print and file-alpha) where the alpha file merge into file-print in the alpha channel. The result is "file-output.tiff"

I hope that is clear now. Thank you in advance.

Best,
Francesco






The tiff

**Reply by 3donew** [reply]
> Hi @fmw42
> 
> My OS is Mac 12.3.1 and ImageMagick 7.1.0-53 Q16-HDRI aarch64 20646
> 
> Yes I attached you the two tiffs that I had (file-print.tiff and file-alpha.tiff) and the result that I want (file-output.tiff) here ( https://www.dropbox.com/t/Hg1O94nTO6mU7XkG )
> 
> I want to merge the two (file-print and file-alpha) where the alpha file merge into file-print in the alpha channel. The result is "file-output.tiff"
> 
> I hope that is clear now. Thank you in advance.
> 
> Best, Francesco
> 
> The tiff

If you find a solution for this problem, please tell me.

**Reply by fmw42** [reply]
You can do that with Imagemagick 7 as follows:

`magick file-print.tif file-alpha.tif -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified file-fred-result.tif`

[file-fred-result.tif.zip](https://github.com/ImageMagick/ImageMagick/files/10187809/file-fred-result.tif.zip)

**Reply by 3donew** [reply]
I have error. No work this command.
magick: unable to open image 'Image.tif': No such file or directory @ error/blob.c/OpenBlob/3570.

**Reply by cesqo** [reply]
> You can do that with Imagemagick 7 as follows:
> 
> `magick file-print.tif file-alpha.tif -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified file-fred-result.tif`
> 
> [file-fred-result.tif.zip](https://github.com/ImageMagick/ImageMagick/files/10187809/file-fred-result.tif.zip)

Hi Fred, it seems great! I'll try tomorrow. I really appreciate your help!

Best,
Francesco

**Comment by 3donew** [comment]
I need to do this too. Here are my files
[https://dropmefiles.com/5vWxC](url)
OS/Windows 11, ImageMagick-7.1.0-47-Q16-HDRI-x64-dll

Thanks

**Reply by fmw42** [reply]
Your link is not valid. Please check. Also see the answer above.

**Reply by 3donew** [reply]
[https://dropmefiles.com/fYqQ8](url)
Check out my files. I am receiving a pdf, the output to the printer should be Tiff. Look at the finished Tif and you can see that the background should be Gray 40-43%. Can this be reproduced in ImageMagik?
Thanks

**Reply by fmw42** [reply]
Link is not valid.  I get a 404 Error. Please fix

**Reply by fmw42** [reply]
Please do not tack a new issue onto someone else's issue.  Open a new issue. Also identify your IM version and platform/OS and show your command line.

**Reply by 3donew** [reply]
Windows 11/ImageMagick-7.1.0-47-Q16-HDRI-x64-dll

**Reply by 3donew** [reply]
New files
https://www.dropbox.com/s/sg4d9p5551opqf6/Files.rar?dl=0

**Reply by 3donew** [reply]
![Cli Error](https://user-images.githubusercontent.com/99094434/206526965-ccc7f857-04ca-4a5e-aedb-b2d59b87b024.png)
Error this command: magick file-print.tif file-alpha.tif -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified file-fred-result.tif

**Reply by 3donew** [reply]
sorry this commad works

**Reply by 3donew** [reply]
![Bad Image](https://user-images.githubusercontent.com/99094434/206531030-bf6ebb0e-2f17-4bd6-a3ef-efbfe2374527.png)

**Reply by fmw42** [reply]
Please explain exactly what you want to do, if you still need help

**Comment by 3donew** [comment]
New file
https://www.dropbox.com/s/limws6vfl6nb3f2/Files.zip?dl=0
https://www.dropbox.com/scl/fo/zsolztailqii0nw6sl0bc/h?dl=0&rlkey=axduyunnji7c4yv1j9h8hkirh

2 pdf files get into our system. 1 picture file, 2 VARNISH file. in order for the printer to start printing, you need to combine them, but the background in the finished picture is gray 40%.
Is it possible to get through ImageMagick the result like in my OutPut Print.tif file using the incoming pdf files?
Or how to do it with tif files?

**Reply by fmw42** [reply]
Does this do what you want?

`magick Input-Image.pdf[0] -colorspace gray Input-VARNISH.pdf -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified output-fred-result.tif`

[output-fred-result.tif.zip](https://github.com/ImageMagick/ImageMagick/files/10188544/output-fred-result.tif.zip)

**Reply by 3donew** [reply]
I checked your CLI, the picture is bad
![bad image](https://user-images.githubusercontent.com/99094434/206541193-c1be9d58-8888-454f-a4c7-38c6630cfa01.png)

why does your files show the correct picture, when I do it I don't have the correct picture? ((
![bad image](https://user-images.githubusercontent.com/99094434/206540859-90868249-cc0f-42de-b741-14158079653f.png)

**Reply by fmw42** [reply]
What is your version of Imagemagick and Ghostscript and libtiff?  And also Photoshop version

**Reply by fmw42** [reply]
I was using IM 7.1.0.53 Q16 Mac OSX Monterey with Ghostscript 9.56 and libtiff 4.4.0 and Photoshop 2022 v24.0.1

**Reply by 3donew** [reply]
![bad image1](https://user-images.githubusercontent.com/99094434/206542296-d500040b-e0c2-4327-972f-a4d4ae9f7971.png)
Imagemagick: ImageMagick-7.1.0-47-Q16-HDRI-x64-dll.exe
Ghostscript: gs9561w64.exe
Photoshop: 2022

I don't have libtiff. where to download it and how to install it?

**Reply by fmw42** [reply]
You need to be using IM 7.1.0.53 There was a bug in -channel-fx that was fixed

**Reply by fmw42** [reply]
You might have it.  What do you get from 

magick -version

I need to see your list of delegates.  Does it include tiff?

**Reply by fmw42** [reply]
But you need to upgrade Imagemagick to 7.1.0.53

**Reply by 3donew** [reply]
![cli](https://user-images.githubusercontent.com/99094434/206543860-ad5241a7-95d0-47e0-8368-54fd12af9c90.png)

**Reply by 3donew** [reply]
I have 3 more questions for you
1. Can I change the resolution of a tif file to 360 dpi?
2. Can the Alpha 1 channel be always on, now when you open tif it is off?
3. Can I change the background so that it is not white but Gray 40% (as in the picture)?

Thank you very much for your help
![ChangeImage40](https://user-images.githubusercontent.com/99094434/206550545-9875311b-5d0f-4a29-bbed-bf102803ca24.png)

**Reply by fmw42** [reply]
> I have 3 more questions for you
> 
> Can I change the resolution of a tif file to 360 dpi?
> Can the Alpha 1 channel be always on, now when you open tif it is off?
> Can I change the background so that it is not white but Gray 40% (as in the picture)?

1) Yes. Add -density 360 just before writing the output

2) I do not know any way to do that with Imagemagick. It might be possible or it might need and enhancement to Imagemgick.  The IM developers would need to comment

3) Does this do what you want?  Unix syntax (Linux or Mac OSX)

`magick \( Input-Image.pdf[0] -colorspace gray -fuzz 15% -fill "gray(40%)" -opaque white \) Input-VARNISH.pdf -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified output-fred-result2.tif`

For Windows, remove the \ from before the opening and closing parenthesis.

[output-fred-result2.tif.zip](https://github.com/ImageMagick/ImageMagick/files/10188995/output-fred-result2.tif.zip)

**Reply by 3donew** [reply]
![density](https://user-images.githubusercontent.com/99094434/206556312-533a5398-c0b8-4c44-8c11-794f06245d02.png)
resolution remains 72dpi

**Reply by fmw42** [reply]
Add -units pixelsperinch also

`magick \( Input-Image.pdf[0] -colorspace gray -fuzz 15% -fill "gray(40%)" -opaque white \) Input-VARNISH.pdf -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified -units pixelsperinch -density 360 output-fred-result2.tif`

[output-fred-result2.tif.zip](https://github.com/ImageMagick/ImageMagick/files/10189174/output-fred-result2.tif.zip)

**Reply by 3donew** [reply]
Thanks a lot. everything is working.
Last question.
You can rename Alpha 1 channel to VARNISH

**Reply by fmw42** [reply]
I do not know how to rename an alpha channel. I do not know that it is possible currently.

**Reply by 3donee** [reply]
Thank you, you helped me a lot

**Reply by 3donew** [reply]
hi fred.
Is there a way to improve the tif image?
After converting from PDF, the picture is very bad, the labels are not readable

**Comment by fmw42** [comment]
The process is to rasterize the PDF with a higher density and then resize by the inverse fraction. Nominal density is 72. So if we use a density of 72*4 = 288 and then resize by 1/4 = 25% you will get higher quality. It will take longer however.

`magick \( -density 288 Input-Image.pdf[0] -resize 25% -colorspace gray -fuzz 15% -fill "gray(40%)" -opaque white \) Input-VARNISH.pdf -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified -units pixelsperinch -density 360 output-fred-result2.tif`

**Reply by 3donew** [reply]
Thanks a lot. It works

**Reply by 3donew** [reply]
Can you tell me how I can smooth out the lines a bit?

**Reply by fmw42** [reply]
Do the same with Input-VARNISH.pdf.

`magick \( -density 288 Input-Image.pdf[0] -resize 25% -colorspace gray -fuzz 15% -fill "gray(40%)" -opaque white \) -density 288 Input-VARNISH.pdf -resize 25% -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified -units pixelsperinch -density 360 output-fred-result2.tif`

Does that help?

**Reply by 3donee** [reply]
Sorry for the long answer, but this team doesn't work for me. When I'm at the computer, I'll show you Image

**Reply by 3donew** [reply]
![Screenshot 2022-12-13 212701](https://user-images.githubusercontent.com/99094434/207415309-d2a2d95a-2282-4ec5-b989-b41d872157cb.png)

**Reply by 3donew** [reply]
The image size is not the same.

**Reply by fmw42** [reply]
Sorry, there should be parentheses around the processing of the second image. So

`magick \( -density 288 Input-Image.pdf[0] -resize 25% -colorspace gray -fuzz 15% -fill "gray(40%)" -opaque white \) \( -density 288 Input-VARNISH.pdf -resize 25% \) -channel-fx "| gray=>alpha" -define tiff:alpha=unspecified -units pixelsperinch -density 360 output-fred-result2.tif`

[output-fred-result2.tif.zip](https://github.com/ImageMagick/ImageMagick/files/10221211/output-fred-result2.tif.zip)

**Reply by 3donew** [reply]
![Screenshot 2022-12-13 224452](https://user-images.githubusercontent.com/99094434/207429803-133f6fc9-960a-4670-9012-258bb80ebc00.png)
alpha filled the entire area and the size of the image should not be the same 350x500mm

**Reply by fmw42** [reply]
That is the natural size of both images.  What do you want? Please explain clearly or show me an example or diagram

**Reply by 3donew** [reply]
![1](https://user-images.githubusercontent.com/99094434/207433556-55809351-a90d-4164-8519-cc22219da037.png)
![2](https://user-images.githubusercontent.com/99094434/207433560-58dbd010-3b44-4209-bf25-52d463d36289.png)

**Reply by fmw42** [reply]
What size for the image?  Should it be the same but just white border?

**Comment by 3donew** [comment]
thise image size 350x500 mm
should have a white border

**Reply by fmw42** [reply]
Is this what you want?  Input-VARNISH has two pages and you only need the first. Also the alpha channel has to be disabled on it.

```
magick \
\( -density 288 Input-Image.pdf[0] -resize 25% -colorspace gray -fuzz 15% -fill "gray(40%)" -opaque white \) \
\( -density 288 Input-VARNISH.pdf[0] -alpha off -resize 25% +write x.png \) \
-channel-fx "| gray=>alpha" \
-define tiff:alpha=unspecified \
-units pixelsperinch -density 360 \
output-fred-result2.tif
```

[output-fred-result2.tif.zip](https://github.com/ImageMagick/ImageMagick/files/10222724/output-fred-result2.tif.zip)

**Reply by 3donew** [reply]
Thanks, it worked.
Tell me, is there a team that makes thin lines a little thicker?

**Reply by fmw42** [reply]
-morphology dilate works for white lines and -morphology erode works for black lines.  See https://imagemagick.org/Usage/morphology/

What do you mean by "team"?

**Reply by 3donew** [reply]
Thanks
